runtime/compiler/control/JITaaSCompilationThread.cpp

```diff
@@ -2363,8 +2349,230 @@ remoteCompile(
+               }
+               break;
+            case J9SHR_RESOURCE_STORE_ERROR:
+               {
+               if (jitConfig->javaVM->sharedClassConfig->verboseFlags & J9SHR_VERBOSEFLAG_ENABLE_VERBOSE)
+                  j9nls_printf( PORTLIB, J9NLS_WARNING,  J9NLS_RELOCATABLE_CODE_STORE_ERROR);
+               TR_J9SharedCache::setSharedCacheDisabledReason(TR_J9SharedCache::SHARED_CACHE_STORE_ERROR);
+               TR::Options::getAOTCmdLineOptions()->setOption(TR_NoLoadAOT);
+               TR::CompilationInfo::disableAOTCompilations();
+               }
+            }
+         }
+      else
+         {
+         if (TR::Options::getAOTCmdLineOptions()->getVerboseOption(TR_VerboseJITaaS))
+            {
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, " Failed AOT cache validation");
+            }
+
+         TR::CompilationInfo::disableAOTCompilations();
+         }
+
+#if defined(J9VM_INTERP_AOT_RUNTIME_SUPPORT)
+
+      TR_Debug *debug = TR::Options::getDebug();
+      bool canRelocateMethod = false;
+      if (debug)
+         {
+         TR_FilterBST *filter = NULL;
+         J9UTF8 *className = ((TR_ResolvedJ9Method*)comp->getCurrentMethod())->_className;
+         J9UTF8 *name = ((TR_ResolvedJ9Method*)comp->getCurrentMethod())->_name;
+         J9UTF8 *signature = ((TR_ResolvedJ9Method*)comp->getCurrentMethod())->_signature;
+         char *methodSignature;
+         char arr[1024];
+         int32_t len = J9UTF8_LENGTH(className) + J9UTF8_LENGTH(name) + J9UTF8_LENGTH(signature) + 3;
+         if (len < 1024)
+            methodSignature = arr;
+         else
+            methodSignature = (char *) TR_MemoryBase::jitPersistentAlloc(len);
+
+         if (methodSignature)
+            {
+            sprintf(methodSignature, "%.*s.%.*s%.*s", J9UTF8_LENGTH(className), utf8Data(className), J9UTF8_LENGTH(name), utf8Data(name), J9UTF8_LENGTH(signature), utf8Data(signature));
+            //printf("methodSig: %s\n", methodSignature);
+
+            if (debug->methodSigCanBeRelocated(methodSignature, filter))
+               canRelocateMethod = true;
+            }
+         else
+            canRelocateMethod = true;
+
+         if (methodSignature && (len >= 1024))
+            TR_MemoryBase::jitPersistentFree(methodSignature);
+         }
+      else
+         {
+         // Prevent the relocation if specific option is given
+         if (!comp->getOption(TR_DisableDelayRelocationForAOTCompilations))
+            canRelocateMethod = false;
+         else
+            canRelocateMethod = true;
+         }
+
+      if (canRelocateMethod)
+         {
+         J9JITDataCacheHeader *cacheEntry;
+
+         TR_ASSERT_FATAL(comp->cg(), "CodeGenerator must be allocated");
+
+         // Use same code cache as AOT compile
+         //
+         TR::CodeCache *aotMCCRuntimeCodeCache = comp->cg()->getCodeCache();
+         TR_ASSERT(aotMCCRuntimeCodeCache, "Must have a reserved codeCache");
```

Comment:
When I tried running with this PR, this assert was triggered.
Normally, JIT would reserve code cache during compilation, but for JITaaS we only reserve the code cache on the server. I'm not sure if it's safe to just remove this assert here.

Reply:
I think we might need to reserve a code cache for JITaaS client remote AOT compilations then.  (If it's used for what I think it's used for, storing the relocated code). I might have disabled it somewhere in this commit. Looking into it.

Reply:
If there isn't a code cache available, allocateSpaceInCodeCache is going to allocate one for us, so I guess it's safe. But I'm not sure if we should still reserve it first. Going to find out.
```
U_8 *
TR_SharedCacheRelocationRuntime::allocateSpaceInCodeCache(UDATA codeSize)
   {
   TR_J9VMBase *fej9 = (TR_J9VMBase *)_fe;
   TR::CodeCacheManager *manager = TR::CodeCacheManager::instance();

   int32_t compThreadID = fej9->getCompThreadIDForVMThread(_currentThread);
   if (!codeCache())
      {
      int32_t numReserved;

      _codeCache = manager->reserveCodeCache(false, codeSize, compThreadID, &numReserved);  // Acquire a cold/warm code cache.
...
```

Reply:
For normal JIT, we reserve the code cache at `performReserveCodeCachePhase` stage. We shouldn't have a code cache at this point for a remote compile. I think we can safely remove the assert.
```
void
OMR::CodeGenPhase::performReserveCodeCachePhase(TR::CodeGenerator * cg, TR::CodeGenPhase * phase)
   {
   cg->reserveCodeCache();
   }
```